### PR TITLE
C#: Refactor `last` predicate

### PR DESCRIPTION
**Pull request by hvitved**
- Split up the `last` predicate into a non-recursive part `lastNonRec` and a recursive part `last`.
- Almost all syntactic constructs have a very simple `last` definition; a set of disjuncts with exactly one recursive call -- those are defined in `lastNonRec`.
- `try` statements and (last) `catch` clauses require multiple recursive calls in the same disjunct, and are therefore handled in the `last` predicate (as before).
- The benefit is that we only need to take care of the join order in the recursive call (for non-`try`/`catch` statements) in one place (the predicate `lastRec`), so we can get rid of many `nomagic`'ed `last`-specialisations.

Profiling report [here](https://git.semmle.com/gist/tom/1bc56f06f5d223fc23d696e5732e0c29) (internal link).

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 1]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
...
             // Operand exits with a Boolean completion
-            exists(BooleanCompletion operandCompletion |
-              result = lastLogicalNotExprOperand(lne, operandCompletion)
-            |
-              c = any(BooleanCompletion bc |
-                  bc.getOuterValue() = operandCompletion.getOuterValue().booleanNot() and
-                  bc.getInnerValue() = operandCompletion.getInnerValue()
-                )
-            )
+            result = lne.getOperand() and
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: I'm not sure how we can deduce that `result = last(result, _)` here?

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 2]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
...
+        TLastRecAbnormalCompletion() or
+        TLastRecBooleanNegationCompletion() or
+        TLastRecNonBooleanCompletion() or
+        TLastRecBreakCompletion() or
+        TLastRecSwitchAbnormalCompletion() or
+        TLastRecInvalidOperationException() or
+        TLastRecNonContinueCompletion() or
+        TLastRecLoopBodyAbnormal()
+
+      private TLastComputation selfValid(ControlFlowElement cfe) {
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Good idea.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 3]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
         }
       }
 
+      /** A specification of how to compute the last element of a control flow element. */
+      private newtype TLastComputation =
+        /** The element is itself the last element. */
+        TSelf(Completion c) or
+        /** The last element must be computed recursively. */
+        TRec(TLastRecComputation c)
```
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: I think I would prefer to keep it as-is, as that gives me exactly the right join-order when using the predicate `lastRec`.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 4]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
...
+                not c instanceof NormalCompletion
+              )
+            )
+          )
       }
 
       pragma[nomagic]
-      private ControlFlowElement lastConditionalQualifiableExprChildExpr(
-        ConditionallyQualifiedExpr cqe, int i, Completion c
+      private ControlFlowElement lastRec(
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: A comment on this predicate would be appreciated.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 5]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
...
-      pragma[nomagic]
-      private ControlFlowElement lastJumpStmtChild(JumpStmt js, Completion c) {
-        result = last(js.getChild(0), c)
-      }
-
-      pragma[nomagic]
-      ControlFlowElement lastTryStmtFinally(TryStmt ts, Completion c) {
-        result = last(ts.getFinally(), c)
+        exists(TLastRecComputation rec |
+          result = lastRec(cfe, c1, rec) and
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Can we simplify this to `result = lastRec(cfe, c1, TLastRecSpecificCompletion(c2))`

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 6]
```diff
@@ -755,588 +755,499 @@ module ControlFlow {
...
       pragma[nomagic]
-      private ControlFlowElement lastArrayCreationInitializer(ArrayCreation ac, Completion c) {
-        result = last(ac.getInitializer(), c)
-      }
-
-      pragma[nomagic]
-      private ControlFlowElement lastArrayCreationLengthArgument(
-        ArrayCreation ac, int i, Completion c
+      private ControlFlowElement lastRecSpecific(
+        ControlFlowElement cfe, Completion c1, Completion c2
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Drop a parameter? This is only ever called with `c1 = c2`

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll (hunk 4 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: We usually don't document folded predicates, but I can add a comment.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll (hunk 5 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Yes.

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll (hunk 6 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: That is by choice, to prevent too eager joining on `c1 = c2`.

**Pull request merged**